text/3373-avoid-nonlocal-definitions-in-fns.md

```diff
@@ -0,0 +1,63 @@
+- Feature Name: N/A
+- Start Date: 2022-01-19
+- RFC PR: [rust-lang/rfcs#3373](https://github.com/rust-lang/rfcs/pull/3373)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Starting in Rust 2024, stop allowing items inside functions to implement
+methods or traits that are visible outside the function.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, tools cross-referencing uses and definitions (such as IDEs) must
+either search inside all function bodies to find potential definitions
+corresponding to uses within a function, or not cross-reference those
+definitions at all.
+
+Humans cross-referencing such uses and definitions may find themselves
+similarly baffled.
+
+With this change, both humans and tools can limit the scope of their search and
+avoid looking for definitions inside other functions, without missing any
+relevant definitions.
+
+# Explanation
+[explanation]: #explanation
+
+Starting in the Rust 2024 edition:
+- An item nested inside a function or closure (through any level of nesting)
+  may not define an `impl Type` block unless the `Type` is also nested inside
+  the same function or closure.
+- An item nested inside a function or closure (through any level of nesting)
+  may not define an `impl Trait for Type` unless either the `Trait` or the
+  `Type` is also nested inside the same function or closure.
```

Comment:
I would say that this should apply to all items with (expression) bodies, that is `function`s, `const`s, `static`s and `enum variant`s. If we only disallow this for functions we still have the same problem with having to look into bodies of other items.

With the exception of `_` unnamed `const` items as those are heavily used by macros to not pollute the surrounding scope with names.

Reply:
Right, those 4 items aren't actually enough either 🙃 

```rust
struct S;
struct Struct<
    const C: usize = {
        impl S {
            const fn method(self) -> usize { 0 }
        }
        0
    },
    const C2: usize = {
        S.method()
    },
>;
```

Reply:
> I would say that this should apply to all items with (expression) bodies, that is `function`s, `const`s, `static`s and `enum variant`s. If we only disallow this for functions we still have the same problem with having to look into bodies of other items.

I see how this applies to `const` and `static`. For enum variants, are you talking about the code on the RHS of an `= const expr`? That's horrifying, good catch.

Reply:
Handled.